### ISSUE TITLE
Handling underflow in kernels

### ISSUE THREAD

**Issue by sdrobert** [SheffieldML/GPyOpt#48]
Hi all,

I've been trying to use your package for hyperparameter optimization. It, like scikit-learn's Gaussian Processes, often underflows when the square exponential kernel is used. If `model_type = 'GP'` and `acquisition_type` is either expected improvement or maximum probability of improvement, I get the error:

```
  File "/a/python3.5/site-packages/GPyOpt/acquisitions/MPI.py", line 38, in _compute_acq
    _, Phi, _ = get_quantiles(self.jitter, fmin, m, s)    
  File "/a/python3.5/site-packages/GPyOpt/util/general.py", line 138, in get_quantiles
    phi = np.exp(-0.5 * u**2) / np.sqrt(2*np.pi)
```

The relevant code snippet is:

``` python
opt_kwargs = {
          'f': self._objective_func,
          'domain': self._space,
          'normalize_Y': False,
          'batch_size': 1,
          'model_type': 'GP',
          'acquisition_type': 'EI',
          'num_cores': 1,
          'exact_feval': False,
          'initial_design_numdata': self.NUM_RANDOM_INIT_POINTS,
}
```

And printing `X`, `Y`, and `space` gives

```
[[0.03278289121021576, 0.9223270601253915, 0.041078090660775515, 1, 1, 1, 1, 2, 1, 0, 0], [0.0459930266775439, 0.5466045681486984, 0.006978856119194612, 0, 0, 0, 2, 2, 1, 2, 1], [0.04378915391665336, 0.578530090404731, 0.09136408190257667, 0, 1, 0, 0, 0, 1, 0, 0], [0.01803270069606775, 0.6932997271162962, 0.08527496299747057, 0, 1, 1, 0, 2, 0, 2, 1], [0.027558962781239666, 0.8592069274365637, 0.034190797138337045, 0, 1, 1, 2, 1, 0, 1, 0], [0.015420387491286701, 0.5956928159271585, 0.04633636991599571, 0, 1, 0, 0, 1, 1, 2, 0], [0.00925519288002172, 0.604910074085572, 0.08911441982609125, 0, 1, 1, 1, 1, 0, 1, 1], [0.04860391821550771, 0.6562647739938442, 0.004181031065854369, 1, 1, 1, 0, 0, 1, 2, 1], [0.012232522924794003, 0.8889700787725456, 0.06011932984882709, 0, 0, 1, 2, 0, 1, 2, 0], [0.007140128397424967, 0.7054756177189522, 0.022160119641586238, 1, 0, 1, 2, 0, 1, 2, 0], [0.003325095442111388, 0.7258961266957953, 0.09103359760065749, 1, 1, 0, 2, 0, 0, 0, 1], [0.021410765938286835, 0.7598611438932474, 0.07642207817613791, 1, 0, 1, 0, 0, 0, 0, 1]] 
[[-0.09225001256533977], [-0.6108390574677939], [-0.49642830217129075], [-0.14971756743156198], [-0.07259310916767191], [-0.6389030419685992], [-0.1849272466827503], [-0.8354279891304349], [-0.07268078200483091], [-0.07268078200483091], [-0.46476992862887007], [-0.07259310916767191]] 
[{'name': 'penalty', 'type': 'continuous', 'domain': (5e-08, 0.05)}, {'name': 'rate_decay', 'type': 'continuous', 'domain': (0.5, 0.99)}, {'name': 'learning_rate', 'type': 'continuous', 'domain': (0.001, 0.1)}, {'name': 'contraction_type', 'type': 'discrete', 'domain': (0, 1)}, {'name': 'weight_init', 'type': 'discrete', 'domain': (0, 1)}, {'name': 'optimizer', 'type': 'discrete', 'domain': (0, 1)}, {'name': 'log2_layer_seq', 'type': 'discrete', 'domain': (0, 1, 2)}, {'name': 'log2_init_fmap', 'type': 'discrete', 'domain': (0, 1, 2)}, {'name': 'log2_fmap_mult', 'type': 'discrete', 'domain': (0, 1)}, {'name': 'log2_batch_size', 'type': 'discrete', 'domain': (0, 1, 2)}, {'name': 'weight_reg', 'type': 'discrete', 'domain': (0, 1)}]
```

Scaling all values between 0 and 1 doesn't help because of the variance normalization term in the kernel. Perhaps kernel calculations would be more stable in the log space? Anyway, it's not clear how to address this.

Thank you for your time,
Sean

**Comment by javiergonzalezh** [comment]
Hi Sean,

I can't see what's is the problem with the info that you have provided. Could you copy and paste the full error message?

Javier

**Comment by sdrobert** [comment]
Hi Javier,

Sorry, it's been a while since I ran this experiment. I remember it had something to do with outliers varying far more than the standard deviation making the exponent very large. If I come across it again, I can give a more detailed account.

Thanks for your time,
Sean
